File: 203_remove_element_from_LL/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution(object):
    def removeElements(self, head, val):
        """
        Remove all nodes with a given value from the linked list.
        
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """
        # Initialize pointers
        curr = head
        dummy = ListNode(0, head)
        prev = dummy

        # Iterate through the linked list
        while curr:
            # Check if the current node's value matches the target value
            if curr.val == val:
                # Remove the current node by updating the previous node's next pointer
                prev.next = curr.next
            else:
                # Move the previous pointer to the current node
                prev = curr
            
            # Move to the next node
            curr = curr.next

        # Return the modified linked list
        return dummy.next

File: 203_remove_element_from_LL/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    /**
     * Remove all nodes with a given value from the linked list.
     * @param head The head of the linked list.
     * @param val The value to be removed.
     * @return The head of the modified linked list.
     */
    ListNode* removeElements(ListNode* head, int val) {
        // Initialize pointers
        ListNode* curr = head;
        ListNode* dummy = new ListNode(0, head);
        ListNode* prev = dummy;

        // Iterate through the linked list
        while (curr) {
            // Check if the current node's value matches the target value
            if (curr->val == val) {
                // Remove the current node by updating the previous node's next pointer
                prev->next = curr->next;
            } else {
                // Move the previous pointer to the current node
                prev = curr;
            }

            // Move to the next node
            curr = curr->next;
        }

        // Return the modified linked list
        return dummy->next;
    }
};
